src/maker/server.rs

```diff
@@ -0,0 +1,561 @@
+
+                log::info!("[{}] Maker server address : {}", maker_port, maker_address);
+
+                log::info!(
+                    "[{}] Directory server address : {}",
+                    maker_port,
+                    directory_onion_address
+                );
+
+                (maker_address, directory_onion_address)
+            }
+        }
+    };
+
+    // Keep trying until send is successful.
+    loop {
+        let mut stream = match maker.config.connection_type {
+            ConnectionType::CLEARNET => match TcpStream::connect(&dns_address) {
+                Ok(s) => s,
+                Err(e) => {
+                    log::warn!(
+                        "[{}] TCP connection error with directory, reattempting: {}",
+                        maker_port,
+                        e
+                    );
+                    thread::sleep(Duration::from_secs(maker.config.heart_beat_interval_secs));
+                    continue;
+                }
+            },
+            ConnectionType::TOR => {
+                match Socks5Stream::connect(
+                    format!("127.0.0.1:{}", maker.config.socks_port),
+                    dns_address.as_str(),
+                ) {
+                    Ok(s) => s.into_inner(),
+                    Err(e) => {
+                        log::warn!(
+                            "[{}] TCP connection error with directory, reattempting: {}",
+                            maker_port,
+                            e
+                        );
+                        thread::sleep(Duration::from_secs(maker.config.heart_beat_interval_secs));
+                        continue;
+                    }
+                }
+            }
+        };
+
+        let request_line = format!("POST {}\n", maker_address);
+        if let Err(e) = stream
+            .write_all(request_line.as_bytes())
+            .and_then(|_| stream.flush())
+        {
+            // Error sending the payload, log and retry after waiting
+            log::warn!(
+                "[{}] Failed to send maker address to directory, reattempting: {}",
+                maker_port,
+                e
+            );
+            thread::sleep(Duration::from_secs(maker.config.heart_beat_interval_secs));
+            continue;
+        }
+        // Payload sent successfully, exit the loop
+        log::info!(
+            "[{}] Successfully sent maker address to directory",
+            maker_port
+        );
+        break;
+    }
+
+    Ok((maker_address, tor_handle))
+}
+
+/// Checks if the wallet already have fidelity bonds. if not, create the first fidelity bond.
+fn setup_fidelity_bond(maker: &Arc<Maker>, maker_address: &str) -> Result<(), MakerError> {
+    let highest_index = maker.get_wallet().read()?.get_highest_fidelity_index()?;
+    if let Some(i) = highest_index {
+        let highest_proof = maker
+            .get_wallet()
+            .read()?
+            .generate_fidelity_proof(i, maker_address)?;
```

Comment:
Nope. That's a recipe for deadlocks. It was previously like that. 

Maybe it will work here. But in general, that's a bad pattern for threaded design. Never hold locks any more than you absolutely need to. Inlining all read and write calls makes deadlock debugging much easier later.

If we have code like this anywhere else in the lib, we should change them too to inline read/write locks. 
`let wallet = maker.get_wallet().read()?;` 

Deadlock happens when you hold a lock like that in a variable, forget to drop it, and then it's moved to another thread. That's a debugging nightmare.